components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt

```diff
@@ -0,0 +1,423 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.appservices.fxaclient
+
+import android.util.Log
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import mozilla.appservices.sync15.DeviceType
+
+// This allows `adb logcat | grep fxa_client` to pick up all log items from this component
+internal const val LOG_TAG = "fxa_client"
+
+/**
+ * Firefox account client
+ */
+class FxaClient private constructor(
+    private val inner: FirefoxAccount,
+    private val persistCallback: FxaPersistCallback,
+    private val eventHandler: FxaEventHandler,
+    coroutineContext: CoroutineContext,
+) : AutoCloseable {
+    /**
+     * CoroutineContext context to run async tasks in.
+     *
+     * This is the coroutineContext passed in to the constructor plus a SupervisorJob.  The
+     * SupervisorJob ensures that if one task fails the other still run.
+     *
+     * Because we use a single CoroutineContext, [close] can cancel all active jobs.
+     *
+     * This is public so applications can use the same context to run their jobs in.  For example
+     * the android-components FxaManager uses it for its startup tasks.
+     */
+    public val coroutineContext = coroutineContext + SupervisorJob()
```

Comment:
Someone with Kotlin experience please double check that I'm doing the correct thing here.